### PR TITLE
Consistent equals and hashCode

### PR DESCRIPTION
**Pull request by darxriggs**
## Summary

Consistent `equals` and `hashCode` implementation using Java 7 [Objects](https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html) helper functions. Tests are performed with [EqualsVerifier](http://jqno.nl/equalsverifier/).

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [ ] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [ ] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [ ] I have interactively tested my changes
- [ ] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by darxriggs** [comment]
Hi @MarkEWaite, any thoughts on this?

**Comment by darxriggs** [comment]
@MarkEWaite ping

**Comment by MarkEWaite** [comment]
@darxriggs I won't look at this until I've completed the review of #553.

**Comment by darxriggs** [comment]
I rebased on master and upgraded to equalsverifier 3.0.3.

File: src/main/java/hudson/plugins/git/extensions/impl/SparseCheckoutPath.java [hunk 1]
```diff
@@ -29,18 +30,21 @@ public String getPath() {
 
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
-        if (!(o instanceof SparseCheckoutPath)) return false;
+        if (this == o) {
```
Comment (MarkEWaite) [inline src/main/java/hudson/plugins/git/extensions/impl/SparseCheckoutPath.java]: Please leave the white space unchanged in lines like this one.

File: src/main/java/jenkins/plugins/git/GitRemoteHeadRefAction.java [hunk 2]
```diff
@@ -62,6 +59,4 @@ public String toString() {
                 ", name='" + name + '\'' +
                 '}';
     }
-
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitRemoteHeadRefAction.java]: Please don't make these types of whitespace changes in the pull request.  They risk merge conflicts with other pull requests and complicate code review

File: src/main/java/jenkins/plugins/git/GitSCMSourceDefaults.java [hunk 3]
```diff
@@ -88,7 +88,7 @@ public boolean equals(Object o) {
      */
     @Override
     public int hashCode() {
-        return (includeTags ? 1 : 0);
+        return includeTags ? 1 : 0;
```
Comment (MarkEWaite) [inline src/main/java/jenkins/plugins/git/GitSCMSourceDefaults.java]: Please remove this change.  It is a change in the style of whitespace changes, modifying a line in a way that risks conflicts on other pull requests.

File: src/main/java/hudson/plugins/git/GitChangeSet.java [hunk 4]
```diff
@@ -612,15 +612,22 @@ public EditType getEditType() {
         }
     }
 
-    public int hashCode() {
-        return id != null ? id.hashCode() : super.hashCode();
+    @Override
```
[the remark this answers is not in the document]
Comment (darxriggs) [reply src/main/java/hudson/plugins/git/GitChangeSet.java]: But that's part of the improvement - having `equals` and `hashCode` implementation consistently in exactly this order.

File: src/main/java/hudson/plugins/git/GitChangeSet.java [hunk 5]
```diff
@@ -472,7 +472,7 @@ private void setMail(User user, String csAuthorEmail) throws IOException {
     private boolean hasMail(User user) {
         hudson.tasks.Mailer.UserProperty property = user.getProperty(hudson.tasks.Mailer.UserProperty.class);
         return property != null && property.hasExplicitlyConfiguredAddress();
-	}
+    }
```
[the remark this answers is not in the document]
Comment (darxriggs) [reply src/main/java/hudson/plugins/git/GitChangeSet.java]: done

File: src/main/java/jenkins/plugins/git/GitRemoteHeadRefAction.java (hunk 2 above)
Comment (darxriggs) [reply src/main/java/jenkins/plugins/git/GitRemoteHeadRefAction.java]: done

File: src/main/java/jenkins/plugins/git/GitSCMSourceDefaults.java (hunk 3 above)
Comment (darxriggs) [reply src/main/java/jenkins/plugins/git/GitSCMSourceDefaults.java]: done

**Comment by darxriggs** [comment]
Hi @MarkEWaite, I removed the whitespace changes as requested. Other requested changes I wouldn't want to perform as commented.

Please let me know if there is anything else to do, e.g. if you should rebase on master or not.

**Comment by MarkEWaite** [comment]
Rebase on master branch would be very nice though even that is not required.  I've marked all the other concerns as "Resolved".

**Comment by darxriggs** [comment]
Rebased on master.

**Pull request merged**